# Gathering data

Available
1. twitter_archive_enhanced.csv

Not-available
1. image_predictions.tsv 
2. tweet_json.txt

In [34]:
#importing various libraries

import pandas as pd
import numpy as np 
import tweepy 
import requests
import json 
from tqdm import *


In [35]:
# Reading the twitter_archive_enhanced.csv file into pandas DataFrame

df_archive = pd.read_csv("twitter-archive-enhanced.csv")
df_archive.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

### Downloading .tsv file programmatically 


Provided url- 
https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv



In [3]:
# Download and Import 'image-predictions.tsv' file Programmatically

url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

# Saveing the file
with open('image-predictions.tsv', "wb") as file: 
    file.write(response.content)
    
# Importing the .tsv file into pandas DataFrame
df_images = pd.read_table('image-predictions.tsv', sep="\t")
df_images.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

## Gathering data from Twitter API

Require
1. consumer_key
2. consumer_secret
3. access_token
4. access_secret
5. tweet_id

In [4]:
# Authenticating to twitter API 

import requests
import tweepy 
import json


with open('AuthKey.txt', 'r') as file:
        consumer_key = file.readline()
        consumer_secret = file.readline()
        access_token = file.readline()
        access_secret = file.readline()

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)
print(api)



In [5]:
# Fetching related data from twitter API using tweet_id

'''
from tqdm import * 


tweet_ids = list(df.tweet_id)

tweet_data = {}
invalid_tweet_id = []

for tweet in tqdm(tweet_ids):
    try:
        tweet_status = api.get_status(tweet,
                                      wait_on_rate_limit=True, 
                                      wait_on_rate_limit_notify=True)
        tweet_data[str(tweet)] = tweet_status._json
    except: 
        invalid_tweet_id.append(tweet_ids)
        
'''

'\nfrom tqdm import * \n\n\ntweet_ids = list(df.tweet_id)\n\ntweet_data = {}\ninvalid_tweet_id = []\n\nfor tweet in tqdm(tweet_ids):\n    try:\n        tweet_status = api.get_status(tweet,\n                                      wait_on_rate_limit=True, \n                                      wait_on_rate_limit_notify=True)\n        tweet_data[str(tweet)] = tweet_status._json\n    except: \n        invalid_tweet_id.append(tweet_ids)\n        \n'

In [6]:
# Converting text data to json

'''
import json

with open('tweet_json.txt', 'w') as file:
        file.write(json.dumps(tweet_data, indent=4))
    
'''

"\nimport json\n\nwith open('tweet_json.txt', 'w') as file:\n        file.write(json.dumps(tweet_data, indent=4))\n    \n"

In [7]:
# Reading tweet_json.txt file into pandas DataFrame

tweet_df = pd.read_json('tweet_json.txt',orient='index')

tweet_df.head(2)

contributors  coordinates          created_at  \
1991-02-08 13:48:08.022790149           NaN          NaN 2015-11-15 22:32:08   
1991-02-08 16:08:05.002620928           NaN          NaN 2015-11-15 23:05:30   

                                                                        entities  \
1991-02-08 13:48:08.022790149  {'hashtags': [], 'symbols': [], 'user_mentions...   
1991-02-08 16:08:05.002620928  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                                               extended_entities  \
1991-02-08 13:48:08.022790149  {'media': [{'id': 666020881337073664, 'id_str'...   
1991-02-08 16:08:05.002620928  {'media': [{'id': 666029276303482880, 'id_str'...   

                               favorite_count  favorited  geo  \
1991-02-08 13:48:08.022790149            2563          0  NaN   
1991-02-08 16:08:05.002620928             129          0  NaN   

                                               id              id_str  \
1991-02-08 13:48:08.022790149  666020888022790144  666020888022790144   
1991-02-08 16:08:05.002620928  666029285002620928  666029285002620928   

                                                     ...                          \
1991-02-08 13:48:08.022790149                        ...                           
1991-02-08 16:08:05.002620928                        ...                           

                              quoted_status  quoted_status_id  \
1991-02-08 13:48:08.022790149           NaN               NaN   
1991-02-08 16:08:05.002620928           NaN               NaN   

                               quoted_status_id_str  retweet_count  retweeted  \
1991-02-08 13:48:08.022790149                   NaN            517          0   
1991-02-08 16:08:05.002620928                   NaN             47          0   

                               retweeted_status  \
1991-02-08 13:48:08.022790149               NaN   
1991-02-08 16:08:05.002620928               NaN   

                                                                          source  \
1991-02-08 13:48:08.022790149  <a href="http://twitter.com/download/iphone" r...   
1991-02-08 16:08:05.002620928  <a href="http://twitter.com/download/iphone" r...   

                                                                            text  \
1991-02-08 13:48:08.022790149  Here we have a Japanese Irish Setter. Lost eye...   
1991-02-08 16:08:05.002620928  This is a western brown Mitsubishi terrier. Up...   

                               truncated  \
1991-02-08 13:48:08.022790149          0   
1991-02-08 16:08:05.002620928          0   

                                                                            user  
1991-02-08 13:48:08.022790149  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1991-02-08 16:08:05.002620928  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[2 rows x 30 columns]

In [36]:
# Copying the dataframes inorder to process

df_clean = df_archive.copy()
images_clean = df_images.copy()
tweet_df_clean = tweet_df.copy()

In [37]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [38]:
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [39]:
tweet_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2336 entries, 1991-02-08 13:48:08.022790149 to 1998-04-12 22:37:23.555336193
Data columns (total 30 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2336 non-null datetime64[ns]
entities                         2336 non-null object
extended_entities                1819 non-null object
favorite_count                   2336 non-null int64
favorited                        2336 non-null int64
geo                              0 non-null float64
id                               2336 non-null int64
id_str                           2336 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null float64
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 non-null float64
is_quote_status                  23

In [40]:
df_clean.index, images_clean.index, tweet_df_clean.index

(RangeIndex(start=0, stop=2356, step=1),
 RangeIndex(start=0, stop=2075, step=1),
 DatetimeIndex(['1991-02-08 13:48:08.022790149',
                '1991-02-08 16:08:05.002620928',
                '1991-02-08 17:16:52.701032449',
                '1991-02-08 20:17:06.329800704',
                '1991-02-08 21:40:48.165822465',
                '1991-02-08 22:05:58.794694657',
                '1991-02-08 22:24:13.826850816',
                '1991-02-08 23:25:25.042405380',
                '1991-02-08 23:51:30.499244032',
                '1991-02-09 00:16:40.524156928',
                ...
                '1998-03-18 16:57:35.349198849',
                '1998-03-22 23:26:25.150312448',
                '1998-03-24 08:46:21.411237888',
                '1998-03-27 04:11:53.173991426',
                '1998-03-28 12:25:50.875897856',
                '1998-03-31 06:59:18.926688256',
                '1998-04-04 11:32:37.279858688',
                '1998-04-05 22:26:21.378084864',
                

In [41]:
df_clean.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [42]:
df_clean[df_clean.duplicated()==True]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [54]:
images_clean[images_clean.duplicated()==True]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

## Assessing data

### Quality issues

1. Could not fetch data from Twitter API for some tweet_ids
2. Type of 'timestamp' column in df_clean is Object instead of timestamp
3. Possible incorrect data in 'name' column like- 'a', 'the', 'this'
4. 'numerator' and 'denominator' columns have some unusual values
5. column name 'id' in tweet_df_clean DataFrame should be changed to 'tweet_id' in-order to merge with other DataFrame
6. Missing values in images_clean DataFrame (2075 rows instead of 2356)
7. We have null values in various column
8. Type of Various column like in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id in tweet_df_clean DataFrame is Float instead of Int
9. Some tweet_ids does not actually contain a picture of dog like- (666052000000000000,666412000000000000)


### Tidiness issues

1. Various unwanted columns in df_clean, and tweet_df_clean DataFrame
2. We can have one column 'dog_stage' instead of doggo floofer pupper and puppo 
3. df_clean, images_clean and tweet_df_clean should be merged to one DataFrame



## Define Steps and Cleaning data

1. Convert column 'timestamp' from Object to timestamp
2. Rename all possible incorrect name to NaN
3. Rename column 'id' to 'tweet_id'
4. Drop all unwanted column from all 3 DataFrame
5. Create one column 'dog_stage' instead of doggo floofer pupper and puppo
6. Merge all 3 DataFrame

In [44]:
df_clean.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

### Define Steps 1

1. Convert column 'timestamp' from Object to timestamp

In [49]:
df_clean['timestamp'] = pd.to_datetime(df_clean['timestamp'])

#df_clean['date'] = df_clean['timestamp'].dt.date
#df_clean['time'] = df_clean['timestamp'].dt.time

### Test

In [52]:
df_clean.timestamp.dtype

dtype('<M8[ns]')

### Define Steps 2

 2. Rename all possible incorrect name to NaN

In [101]:
df_clean.name.replace(to_replace = ['a','all','an','by','his','my','just','this','the'], value=None, inplace=True)

In [67]:
df_clean[df_clean.name == 'a']

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [68]:
df_clean[df_clean.name == 'my']

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

### Define Steps 3

3. Rename column 'id' to 'tweet_id'

In [72]:
tweet_df_clean.rename(columns={'id': 'tweet_id'}, inplace=True)

### Test

In [73]:
tweet_df_clean.head(1)

contributors  coordinates          created_at  \
1991-02-08 13:48:08.022790149           NaN          NaN 2015-11-15 22:32:08   

                                                                        entities  \
1991-02-08 13:48:08.022790149  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                                               extended_entities  \
1991-02-08 13:48:08.022790149  {'media': [{'id': 666020881337073664, 'id_str'...   

                               favorite_count  favorited  geo  \
1991-02-08 13:48:08.022790149            2563          0  NaN   

                                         tweet_id              id_str  \
1991-02-08 13:48:08.022790149  666020888022790144  666020888022790144   

                                                     ...                          \
1991-02-08 13:48:08.022790149                        ...                           

                              quoted_status  quoted_status_id  \
1991-02-08 13:48:08.022790149           NaN               NaN   

                               quoted_status_id_str  retweet_count  retweeted  \
1991-02-08 13:48:08.022790149                   NaN            517          0   

                               retweeted_status  \
1991-02-08 13:48:08.022790149               NaN   

                                                                          source  \
1991-02-08 13:48:08.022790149  <a href="http://twitter.com/download/iphone" r...   

                                                                            text  \
1991-02-08 13:48:08.022790149  Here we have a Japanese Irish Setter. Lost eye...   

                               truncated  \
1991-02-08 13:48:08.022790149          0   

                                                                            user  
1991-02-08 13:48:08.022790149  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[1 rows x 30 columns]

### Define Steps 4

4. Drop all unwanted column from all 3 DataFrame

In [76]:
df_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'source', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'expanded_urls'], axis=1, inplace = True)

In [80]:
tweet_df_clean.drop(['contributors','coordinates','extended_entities','favorited','geo','id_str','in_reply_to_screen_name','in_reply_to_status_id','in_reply_to_status_id_str','in_reply_to_user_id','in_reply_to_user_id_str','is_quote_status','place','possibly_sensitive','possibly_sensitive_appealable','quoted_status','quoted_status_id','quoted_status_id_str','retweeted','retweeted_status','truncated','user'], axis=1, inplace = True)

### Test

In [77]:
list(df_clean)

['tweet_id',
 'timestamp',
 'text',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

In [78]:
list(images_clean)

['tweet_id',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog']

In [81]:
list(tweet_df_clean)

['created_at',
 'entities',
 'favorite_count',
 'tweet_id',
 'lang',
 'retweet_count',
 'source',
 'text']

### Define Steps 5

5. Create one column 'dog_stage' instead of doggo floofer pupper and puppo

In [82]:
columns_to_melt = ['doggo', 'floofer', 'pupper', 'puppo']
columns_to_stay = [x for x in df_clean.columns.tolist() if x not in columns_to_melt]

# Mlet the the columns into values
df_clean = pd.melt(df_clean, id_vars = columns_to_stay, value_vars = columns_to_melt, 
                         var_name = 'stages', value_name = 'dog_stage')

df_clean.drop(['stages'], axis=1, inplace=True)

### Test

In [86]:
list(df_clean)

['tweet_id',
 'timestamp',
 'text',
 'rating_numerator',
 'rating_denominator',
 'name',
 'dog_stage']

In [87]:
df_clean.dog_stage.value_counts()

None       9030
pupper      257
doggo        97
puppo        30
floofer      10
Name: dog_stage, dtype: int64

In [88]:
df_archive.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

### Define Steps 6

6. Merge all 3 DataFrame

In [102]:
df_arcimg = pd.merge(df_clean, images_clean, how='inner')
df_all = pd.merge(df_arcimg, tweet_df_clean, how='inner')

### Test

In [103]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4812 entries, 0 to 4811
Data columns (total 24 columns):
tweet_id              4812 non-null int64
timestamp             4812 non-null datetime64[ns]
text                  4812 non-null object
rating_numerator      4812 non-null int64
rating_denominator    4812 non-null int64
name                  4812 non-null object
dog_stage             4812 non-null object
jpg_url               4812 non-null object
img_num               4812 non-null int64
p1                    4812 non-null object
p1_conf               4812 non-null float64
p1_dog                4812 non-null bool
p2                    4812 non-null object
p2_conf               4812 non-null float64
p2_dog                4812 non-null bool
p3                    4812 non-null object
p3_conf               4812 non-null float64
p3_dog                4812 non-null bool
created_at            4812 non-null datetime64[ns]
entities              4812 non-null object
favorite_count        48

In [104]:
# Saving the final DataFrame

df_all.to_csv('df_final.csv')